core/node/consensus/src/storage/tests.rs

```diff
@@ -0,0 +1,92 @@
+use rand::Rng;
+use zksync_basic_types::web3::contract::Tokenize;
+use zksync_concurrency::{ctx, scope};
+use zksync_types::{
+    api::{BlockId, BlockNumber},
+    ethabi::{Address, Token},
+    L2ChainId, ProtocolVersionId, U256,
+};
+
+use crate::storage::ConnectionPool;
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_vm_reader() {
+    zksync_concurrency::testonly::abort_on_panic();
+    let ctx = &ctx::test_root(&ctx::RealClock);
+    let rng = &mut ctx.rng();
+
+    scope::run!(ctx, |ctx, s| async {
+        let pool = ConnectionPool::test(false, ProtocolVersionId::latest()).await;
+        let (node, runner) = crate::testonly::StateKeeper::new(ctx, pool.clone()).await?;
+        let account = runner.account.clone();
+        s.spawn_bg(runner.run_real(ctx));
+
+        let mut writer = super::testonly::VMWriter::new(pool.clone(), node, account.clone());
+
+        let mut nodes: Vec<Vec<Token>> = Vec::new();
+        let num_nodes = 5;
+        for _ in 0..num_nodes {
+            let node_entry = (
+                Address::random(),
+                U256::from(rng.gen::<usize>()),
+                (0..256).map(|_| rng.gen()).collect::<Vec<u8>>(),
+                (0..256).map(|_| rng.gen()).collect::<Vec<u8>>(),
+                U256::from(rng.gen::<usize>()),
+                (0..256).map(|_| rng.gen()).collect::<Vec<u8>>(),
+            )
+                .into_tokens();
```

Comment:
Maybe you could use [abigen!](https://docs.rs/ethers/latest/ethers/contract/macro.abigen.html) to generate bindings for all the Solidity structs and then you get the tokenization with static typing.

Reply:
We didn't migrate to `ethers` just yet (and in fact it's already deprecated in favor of `alloy`), so using it is not recommended for now. We want the migration to be done consistently, otherwise, we'll end up with a mix of our vendored `web3` remains and some of the newer stuff.